### PR TITLE
Implement np.linalg.matrix_power

### PR DESCRIPTION
**Pull request by ziyadedher**
Implements [`np.linalg.matrix_rank`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.matrix_power.html).

Implementation is heavily based on [the NumPy implementation](https://github.com/numpy/numpy/blob/v1.17.0/numpy/linalg/linalg.py#L559-L672) with some tweaks to ensure JAX and JIT compatibility.

Related to #1999

### CODE REVIEW

**Review by shoyer — COMMENTED** [review]
Looks very nice!

I wonder -- would it to be a good idea to write a custom JVP rule for `matrix_power` that implements the gradient directly rather than by differentiation the implementation?

Either way, this looks good for now. It's probably best to hold off on any custom JVPs until after https://github.com/google/jax/pull/2026 goes in.

File: jax/numpy/linalg.py [hunk 1]
```diff
@@ -62,6 +63,53 @@ def svd(a, full_matrices=True, compute_uv=True):
...
+    fmatmul = np.matmul
+  elif a.ndim == 2:
+    fmatmul = np.dot
+  else:
+    raise NotImplementedError(
+        "matrix_power not supported for stacks of object arrays")
+
+  if n == 0:
+    a = np.empty_like(a)
+    a = ops.index_update(a, ..., np.eye(a.shape[-2], dtype=a.dtype))
```
[the remark this answers is not in the document]
Comment (ziyadedher) [reply jax/numpy/linalg.py]: If we have that the original `a` has shape like `(i, j, k, k)` (notice that the final two dimension sizes must be equal), then what we are returning here is essentially an `i` by `j` matrix of `k` by `k` identity matrices. In the case of 2D matrix power, this is essentially equivalent to what you mentioned; however, in higher dimensions (where honestly I cannot wrap my head around what matrix power is actually doing, intuitively) this differs.

I tried thinking of ways to get around this, since these couple lines do bring about this question that you asked at first glance, but nothing came to mind. Any thoughts?

**Comment by ziyadedher** [comment]
@shoyer I think a custom JVP rule would definitely be useful given the current implementation of `matrix_power`. I'll try to monitor #2026 and implement the JVP once it's resolved.

**Comment by shoyer** [comment]
OK, last request -- please this function to the list in `docs/jax.numpy.rst` so we have a docstring!

**Comment by ziyadedher** [comment]
@shoyer Done!

**Comment by shoyer** [comment]
Thank you @ziyadedher !
